### DISCUSSION TITLE
Content Items event handler, UpdatingAsync method: property "part" and "context.UpdatingItem.Elements" has always null properties

### DISCUSSION THREAD

**Discussion by MarGraz** [OrchardCMS/OrchardCore#15952] [Q&A] [answered]
Hi everyone,

I already discussed a similar topic [here](https://github.com/OrchardCMS/OrchardCore/discussions/15568). Last time, when I used the `UpdatingAsync` method, I was able to achieve my goal.

Now I have implemented a module with a `CustomBagPartHandler : ContentPartHandler<BagPart>`, which I intend to use to **validate the number of items passed inside a** `BagPart` **contained within a new** `ContentType` (e.g. a `User`).

Here is my handler implementation:

```csharp
    public class CustomBagPartHandler : ContentPartHandler<BagPart>
    {
        private readonly IContentDefinitionManager _contentDefinitionManager;
        private readonly IActionContextAccessor _actionContextAccessor;

        public CustomBagPartHandler(IContentDefinitionManager contentDefinitionManager, 
            IActionContextAccessor actionContextAccessor)
        {
            _contentDefinitionManager = contentDefinitionManager;
            _actionContextAccessor = actionContextAccessor;
        }

        public override async Task UpdatingAsync(UpdateContentContext context, BagPart part)
        {
            BagPartCustomSettingsList? settings = await GetSettingsAsync(part);

            if (settings?.ContentTypeSettings != null)
            {
                foreach (BagPartCustomContentTypeSettings contentTypeSetting in settings.ContentTypeSettings)
                {
                    // Filter content items by the current content type setting
                    IEnumerable<ContentItem> contentItemsOfType = part.ContentItems.Where(ci => ci.ContentType == contentTypeSetting.ContentType);
                    int count = contentItemsOfType.Count();

                    // Validate against the min and max settings for this content type
                    // Allow the possibility to have Min and Max = 0
                    if (count < contentTypeSetting.MinItems || (contentTypeSetting.MaxItems > 0 && count > contentTypeSetting.MaxItems))
                    {
                        _actionContextAccessor.ActionContext?.ModelState?.AddModelError(nameof(CustomBagPartHandler), $"The number of {contentTypeSetting.ContentType} items in the bag must be between {contentTypeSetting.MinItems} and {contentTypeSetting.MaxItems}. Current count: {contentItemsOfType.Count()}.");
                    }
                }
            }

            await Task.CompletedTask;
        }

        private async Task<BagPartCustomSettingsList> GetSettingsAsync(BagPart part)
        {
            ContentTypeDefinition contentTypeDefinition = await _contentDefinitionManager.GetTypeDefinitionAsync(part.ContentItem.ContentType); 
            ContentTypePartDefinition? contentTypePartDefinition = contentTypeDefinition?.Parts?.FirstOrDefault(x => string.Equals(x.PartDefinition.Name, nameof(BagPart), StringComparison.Ordinal));

            return contentTypePartDefinition?.GetSettings<BagPartCustomSettingsList>();
        }
    }
```

The handler correctly retrieves the `BagPartCustomSettingsList`, but **it seems impossible to access the** `User` **fields and parts**. Exploring `context.UpdatingItem.Content` or `part`, the object appears to have all null values.

Here is **what's inside** `context.UpdatingItem.Content`:

![327532472-76481f06-dfeb-4c66-bd51-97e905811bf8](https://github.com/OrchardCMS/OrchardCore/assets/19241483/c7aad48d-1e8b-46bb-a5f7-0191bae92f5a)

And here is **what's inside** `part`: 

![327532917-84a92d93-ffa2-4c9e-ad7e-e2ab1e84d6b6](https://github.com/OrchardCMS/OrchardCore/assets/19241483/46b67d48-1202-406b-9345-b39894b3aa20)

I suspected that during the `UpdatingAsync`, the model is not yet bound, so I attempted to use the `ValidatingAsync` and `ValidatedAsync` methods, **but they are never triggered** 🤯🙃

This is how I implemented the `ValidatingAsync` method:

```csharp
public override async Task ValidatingAsync(ValidateContentContext context, BagPart part)
{
    var settings = await GetSettingsAsync(part);

    if (settings?.ContentTypeSettings != null)
    {
        foreach (var setting in settings.ContentTypeSettings)
        {
            var contentItemsOfType = part.ContentItems.Where(ci => ci.ContentType == setting.ContentType);
            int count = contentItemsOfType.Count();

            if (count < setting.MinItems || (setting.MaxItems > 0 && count > setting.MaxItems))
            {
                context.Fail(string.Format("The number of {0} items in the bag must be between {1} and {2}. Current count: {3}.",
                    setting.ContentType, setting.MinItems, setting.MaxItems, count));
            }
        }
    }
}
```

Unfortunately the `IContentPartHandler` interface ([here](https://github.com/OrchardCMS/OrchardCore/blob/main/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/Handlers/IContentPartHandler.cs)) doesn't have any `summary` for each method, which could have offered some guidance.

What am I doing wrong? I have spent several hours trying to fix this, also searching for an example in the Orchard Core solution, without success. 

Thank you

**Comment by Piedone** [comment]
I don't have a solution from the top of my head, but here's a video with voice about how I'd approach figuring it out.

https://github.com/OrchardCMS/OrchardCore/assets/1976647/a19a0b98-7741-484e-b2ef-02645b6b71c4

**Reply by MarGraz** [reply]
@Piedone, thank you for your reply and the video explanation 😊

I can't use a breakpoint in the `DefaultContentManager` class because it's part of OC NuGet. However, opening the Orchard Core solution helped me better understand how it works. I always do this for inspiration 🤓

**Regarding the** `IContentManager` **interface**, I think I need to focus on the `New` and `Update` methods. These involve **published and draft items**, so methods like:

- `NewAsync`
- `UpdateAsync`
- `CreateAsync`
- `CreateContentItemVersionAsync`
- `UpdateContentItemVersionAsync` 

are crucial. They are visible [here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement.Abstractions/IContentManager.cs#L15C8-L51C124).

In the `DefaultContentManager` class, methods like `NewAsync` [here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L54-L82), and `UpdateAsync` [here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L775C8-L784C10), don't call `ValidatingAsync`. 🤔

The `ValidateAsync` [method](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L786-L800) (that includes `ValidatingAsync`) is used in other "sub"-methods (including private ones) which handle additional logic. 
This explains why my `ValidatingAsync` and `ValidatedAsync` overrides methods are never triggered.

### What have I done?

**I tried a new approach** using `CreatingAsync` and `UpdatingAsync` methods. They share the same logic, but a different condition to execute the logic. Infact, the `UpdatingAsync` is always executed, both when a Content Item is updated, either when is created, and this sounds logically strange for me.

**In the end, the problem remains the same**: this approach works for newly created Content Items but not for updated ones. The `part` in `UpdatingAsync` **_never shows_** the newly added parts.

Here's the new approach implementation:

```csharp
    public class Startup : StartupBase
    {
        public override void ConfigureServices(IServiceCollection services)
        {    
            // Handlers

            services.AddContentPart<BagPart>().AddHandler<CustomBagPartHandler>();

            // Drivers

            services.AddScoped<IContentTypePartDefinitionDisplayDriver, BagPartCustomSettingsDisplayDriver>();
        }
     }
```

 ```csharp
 public class CustomBagPartHandler : ContentPartHandler<BagPart>
  {
      private readonly IContentDefinitionManager _contentDefinitionManager;
      private readonly IActionContextAccessor _actionContextAccessor;

      public CustomBagPartHandler(IContentDefinitionManager contentDefinitionManager, 
          IActionContextAccessor actionContextAccessor)
      {
          _contentDefinitionManager = contentDefinitionManager;
          _actionContextAccessor = actionContextAccessor;
      }

      public override async Task CreatingAsync(CreateContentContext context, BagPart part) {

          BagPartCustomSettingsList? settings = await GetSettingsAsync(part);

          // Ensure settings are correctly retrieved
          if (settings?.ContentTypeSettings != null)
          {
              ValidateCustomBagPart(part, settings);
          }

          await Task.CompletedTask;
      }

      public override async Task UpdatingAsync(UpdateContentContext context, BagPart part)
      {
          BagPartCustomSettingsList? settings = await GetSettingsAsync(part);

          // Ensure settings are correctly retrieved and that is not a new Content Type creation
          if (settings?.ContentTypeSettings != null && 
                part.ContentItems.Count != 0 && 
                context.UpdatingItem.CreatedUtc != null)
          {
              ValidateCustomBagPart(part, settings);
          }

          await Task.CompletedTask;
      }

      #region Private methods

      private async Task<BagPartCustomSettingsList> GetSettingsAsync(BagPart part)
      {
          ContentTypeDefinition contentTypeDefinition = await _contentDefinitionManager.GetTypeDefinitionAsync(part.ContentItem.ContentType);
          ContentTypePartDefinition? contentTypePartDefinition = contentTypeDefinition?.Parts?.FirstOrDefault(x => string.Equals(x.PartDefinition.Name, nameof(BagPart), StringComparison.Ordinal));

          return contentTypePartDefinition?.GetSettings<BagPartCustomSettingsList>();
      }

      private void ValidateCustomBagPart(BagPart part, BagPartCustomSettingsList settings)
      {
          foreach (BagPartCustomContentTypeSettings contentTypeSetting in settings.ContentTypeSettings)
          {
              // Filter content items by the current content type setting
              IEnumerable<ContentItem> contentItemsOfType = part.ContentItems.Where(ci => ci.ContentType == contentTypeSetting.ContentType);
              int count = contentItemsOfType.Count();

              // Validate against the min and max settings for this content type
              // Allow the possibility to have Min and Max = 0
              if (count < contentTypeSetting.MinItems || (contentTypeSetting.MaxItems > 0 && count > contentTypeSetting.MaxItems))
              {
                  // Return error
                  _actionContextAccessor.ActionContext?.ModelState?.AddModelError(nameof(CustomBagPartHandler), $"The number of {contentTypeSetting.ContentType} items in the bag must be between {contentTypeSetting.MinItems} and {contentTypeSetting.MaxItems}. Current count: {contentItemsOfType.Count()}.");
              }
          }
      }

      #endregion
  }
```

Questions:

1. Why isn't `ValidateAsync` ([here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L786-L800)) done earlier, for example inside `NewAsync` or `CreateAsync`, where it doesn't occur at all, and in the `UpdateAsync` is done deeper in the logic?

2. I noticed that using `UpdatedAsync` (instead of `UpdatingAsync`) allows me to get the `part` non-null, but at that point the Content Item is already saved, so is not anymore a validation.

3. In the new approach I tried above, why is `UpdatingAsync` called during a **creation** of a new Content Item? This logic sounds really strange 🤔

4. Why doesn't `UpdatingAsync` display the new updated bag parts that are coming in the `part`? Neither in the `context.UpdatingItem.Elements`. Could this be a bug? 😟

Thank you

**Reply by MarGraz** [reply]
Is it possible to ask the developer **who created those methods to explain the best practices for using them**? 

I know I might be asking a lot 🫠, but I think it could be useful for everyone to have a clear set of best practices to follow. Then, we could also create a reference or add summaries to the `IContentPartHandler` interface.

Thank you

**Reply by Piedone** [reply]
I'm getting a bit overloaded here. Could you please formulate what you need help with and couldn't figure out from the OC source as as a single question in a paragraph?

**Reply by MarGraz** [reply]
Hi @Piedone,

Yes, I have done a complete overview of what I have done with the encountered issues. Perhaps it's a bit much 😅

### Summary:

In OC 1.8.3, in the `DefaultContentManager` class, methods like `NewAsync` ([here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L54-L82)) and `UpdateAsync` ([here](https://github.com/OrchardCMS/OrchardCore/blob/be675b3e1805f3a7b30afff9ceab747719ca9091/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs#L775C8-L784C10)) do not call `ValidateAsync`. This explain why my handler overrides for `ValidatingAsync` and `ValidatedAsync` are never triggered.

**I have implemented a new approach** (you can find it in the [message above](https://github.com/OrchardCMS/OrchardCore/discussions/15952#discussioncomment-9303986)). Here's what I observed:

- `CreatingAsync` is called only the first time a new Content Item is created, which is the expected behavior.
- `UpdatingAsync` is always called, even when creating a new Content Item. This is strange.
- **When creating** a new Content Item, in the `UpdatingAsync` method I receive a **null** value `part` and `context.UpdatingItem.Elements` (as shown in the screenshots in the [first message](https://github.com/OrchardCMS/OrchardCore/discussions/15952#discussion-6597429)). 
Instead, **when updating** an existing Content Item, `UpdatingAsync` retrieves in the `part` and `context.UpdatingItem.Elements` the previously saved version, not the new one I need to validate.

Could this be a bug? 
How can I add a validation handler for my bag part content items?

Thank you

> **Update**: fixed the misused "Content Type", with the correct word "Content Item".

**Reply by Piedone** [reply, marked as answer]
I believe when you say new/existing "Content Type", you mean "content item" instead.

- The Validate* handlers are only called from `ContentManager.ValidateAsync()`. This seems correct.
- `UpdatingAsync` being called for a new content type also seems correct if you check out where it's called.
- `UpdatingAsync` receiving `null` in its `part` parameter for new content items doesn't seem correct. However, that's not what I see in your screenshot. It being empty (but not `null) is expected, since this event is called before the content item is updated with new data from POST. Due to this, it's also expected that you'll see some value (the old one, before the update) for an existing content item.
- You can add custom validation by injecting `IUpdateModelAccessor` in your handler, and adding model state errors to it in `UpdatedAsync`. Orchard Core Commerce [has a lot of examples of this](https://github.com/search?q=repo%3AOrchardCMS%2FOrchardCore.Commerce%20IUpdateModelAccessor&type=code).

**Reply by MarGraz** [reply]
Thank you for your answer, as always 😊  
Yes, I meant "Content Item". I have updated the previous post, thank you for noticing it.

- `UpdatingAsync` being called for the creation of a new content item sounds odd to me, because it's not an "update", but a "create" event 🤔  
When I checked the `DefaultContentManager` in the OC solution, I didn't find `UpdateAsync` being used in the `NewAsync` and `CreateAsync` [methods](https://github.com/OrchardCMS/OrchardCore/blob/release/1.8.3/src/OrchardCore/OrchardCore.ContentManagement/DefaultContentManager.cs). I'm using the OC NuGet package, so it is impossible for me to use breakpoints inside the `DefaultContentManager` (to do that, I would need to replicate the same module and content items in the local OC solution).

- **Injecting `IUpdateModelAccessor` and using `UpdatedAsync` works** 🥳  
It seems that using `UpdatedAsync` does the validation after the first save, and **if the validation fails, the changes are reverted, right?** If that is the case, I wouldn't have written to the database, to perform later a revert. This could save at least two database calls. However, given the architecture of OC, this was evidently the only way to do it (I'm curious).

Here is the updated code:

```csharp
    public class CustomBagPartHandler : ContentPartHandler<BagPart>
    {
        private readonly IUpdateModelAccessor _updateModelAccessor;
        private readonly IContentDefinitionManager _contentDefinitionManager;
        private readonly IStringLocalizer<CustomBagPartHandler> S;

        public CustomBagPartHandler(
            IUpdateModelAccessor updateModelAccessor,
            IContentDefinitionManager contentDefinitionManager,
            IStringLocalizer<CustomBagPartHandler> stringLocalizer)
        {
            _updateModelAccessor = updateModelAccessor;
            _contentDefinitionManager = contentDefinitionManager;
            S = stringLocalizer;
        }

        public override async Task UpdatedAsync(UpdateContentContext context, BagPart part)
        {
            BagPartCustomSettingsList? settings = await GetSettingsAsync(part);

            // Ensure settings are correctly retrieved
            if (settings?.ContentTypeSettings != null)
            {
                foreach (BagPartCustomContentTypeSettings contentTypeSetting in settings.ContentTypeSettings)
                {
                    // Filter content items by the current content type setting
                    IEnumerable<ContentItem> contentItemsOfType = part.ContentItems.Where(ci => ci.ContentType == contentTypeSetting.ContentType);
                    int count = contentItemsOfType.Count();

                    // Validate against the min and max settings for this content type
                    // Allow the possibility to have Min and Max = 0
                    if (count < contentTypeSetting.MinItems || (contentTypeSetting.MaxItems > 0 && count > contentTypeSetting.MaxItems))
                    {
                        // Return error and revert changes
                        _updateModelAccessor.ModelUpdater.ModelState.AddModelError(nameof(CustomBagPartHandler), S["The number of {0} items in the bag must be between {1} and {2}. Current count: {3}. All changes will be ignored", contentTypeSetting.ContentType, contentTypeSetting.MinItems, contentTypeSetting.MaxItems, contentItemsOfType.Count()]);
                        return;
                    }
                }
            }
            await Task.CompletedTask;
        }

        #region Private methods

        private async Task<BagPartCustomSettingsList> GetSettingsAsync(BagPart part)
        {
            ContentTypeDefinition contentTypeDefinition = await _contentDefinitionManager.GetTypeDefinitionAsync(part.ContentItem.ContentType);
            ContentTypePartDefinition? contentTypePartDefinition = contentTypeDefinition?.Parts?.FirstOrDefault(x => string.Equals(x.PartDefinition.Name, nameof(BagPart), StringComparison.Ordinal));

            return contentTypePartDefinition?.GetSettings<BagPartCustomSettingsList>();
        }

        #endregion
    }
```

Thank you

**Reply by Piedone** [reply]
"Update" in this context means updating the content item object from the HTTP request. This happens for freshly created content items too.

Since OC uses SourceLink, you can actually step into it even with NuGet packages, and put breakpoints in OC code. You need to uncheck “Enable Just My Code” in VS for this to work.

Since updating the whole content item is usually a single query, at the end, canceling it in the middle from a handler doesn't cause anything excessive.

**Reply by MarGraz** [reply]
Thank you for the explanation 😊
